### PR TITLE
Significantly reduce allocations in ExpandMetadataLeaveEscaped

### PR DESCRIPTION
**Pull request by Erarndt**
### Context

The current implementation of `ExpandMetadataLeaveEscaped()` heavily leverages Regex methods, and doing so allocates a considerable amount. I measure around 3.8GB of allocations in this path while building Roslyn. Many of these allocations come from a combination of internal Regex implementation and the existing patterns of consuming them. Take this code for example:

                        using SpanBasedStringBuilder finalResultBuilder = Strings.GetSpanBasedStringBuilder();

                        int start = 0;
                        MetadataMatchEvaluator matchEvaluator = new MetadataMatchEvaluator(metadata, options, elementLocation, loggingContext);

                        if (itemVectorExpressions != null)
                        {
                            // Move over the expression, skipping those that have been recognized as an item vector expression
                            // Anything other than an item vector expression we want to expand bare metadata in.
                            for (int n = 0; n < itemVectorExpressions.Count; n++)
                            {
                                string vectorExpression = itemVectorExpressions[n].Value;

                                // Extract the part of the expression that appears before the item vector expression
                                // e.g. the ABC in ABC@(foo->'%(FullPath)')
                                string subExpressionToReplaceIn = expression.Substring(start, itemVectorExpressions[n].Index - start);
                                string replacementResult = RegularExpressions.NonTransformItemMetadataRegex.Replace(subExpressionToReplaceIn, new MatchEvaluator(matchEvaluator.ExpandSingleMetadata));

                                // Append the metadata replacement
                                finalResultBuilder.Append(replacementResult);

                                // Expand any metadata that appears in the item vector expression's separator
                                if (itemVectorExpressions[n].Separator != null)
                                {
                                    vectorExpression = RegularExpressions.NonTransformItemMetadataRegex.Replace(itemVectorExpressions[n].Value, new MatchEvaluator(matchEvaluator.ExpandSingleMetadata), -1, itemVectorExpressions[n].SeparatorStart);
                                }

                                // Append the item vector expression as is
                                // e.g. the @(foo->'%(FullPath)') in ABC@(foo->'%(FullPath)')
                                finalResultBuilder.Append(vectorExpression);

                                // Move onto the next part of the expression that isn't an item vector expression
                                start = (itemVectorExpressions[n].Index + itemVectorExpressions[n].Length);
                            }
                        }

Here there are several calls to `Regex.Replace()` that each return a new string that gets appended to the `SpanBasedStringBuilder`. If you dig into the implementation of `Regex.Replace()` you can start to see where the inefficiencies really add up

            stringBuilder = new StringBuilder();
            int num = 0;
            do
            {
                if (match.Index != num)
                {
                    stringBuilder.Append(input, num, match.Index - num);
                }

                num = match.Index + match.Length;
                stringBuilder.Append(evaluator(match));
                if (--count == 0)
                {
                    break;
                }

                match = match.NextMatch();
            }
            while (match.Success);
            if (num < input.Length)
            {
                stringBuilder.Append(input, num, input.Length - num);
            }

So, for *each* call to `Regex.Replace()` we're, we're creating a new `StringBuilder` and appending to it, causing it to expand it's internal data structures, as we evaluate the matches. Finally, we return the resulting string and add it to the span based builder. By modifying this implementation to directly use the `SpanBasedStringBuilder`, we can eliminate a considerable amount of duplication. We can do some additional cleanup in this code to avoid some additional allocations as well.

Before:

image

After:

image

Many types that were previously allocated are either entirely eliminated or significantly reduced. Before there are ~3.8GB of allocations here and only 2.7GB here after.

### CODE REVIEW

**Comment by SimaTian** [comment]
There is https://github.com/dotnet/msbuild/pull/12009 targetting the same area. Are these complementary or is this one making the other redundant please?

**Comment by Erarndt** [comment]
> There is #12009 targetting the same area. Are these complementary or is this one making the other redundant please?

Chatted offline. The changes are in related but separate areas.

File: src/Build/Evaluation/Expander.cs [hunk 1]
```diff
@@ -985,20 +1000,20 @@ internal static string ExpandMetadataLeaveEscaped(string expression, IMetadataTa
...
                                 if (itemVectorExpressions[n].Separator != null)
                                 {
-                                    vectorExpression = RegularExpressions.NonTransformItemMetadataRegex.Replace(itemVectorExpressions[n].Value, new MatchEvaluator(matchEvaluator.ExpandSingleMetadata), -1, itemVectorExpressions[n].SeparatorStart);
+                                    RegularExpressions.ReplaceAndAppend(itemVectorExpressions[n].Value, MetadataMatchEvaluator.ExpandSingleMetadata, matchEvaluator, -1, itemVectorExpressions[n].SeparatorStart, finalResultBuilder, RegularExpressions.NonTransformItemMetadataRegex);
+                                }
+                                else
+                                {
+                                    // Append the item vector expression as is
+                                    // e.g. the @(foo->'%(FullPath)') in ABC@(foo->'%(FullPath)')
+                                    finalResultBuilder.Append(vectorExpression);
```
Comment (SimaTian) [inline src/Build/Evaluation/Expander.cs]: To make sure I understand correctly: The ReplaceAndAppend now handles the append that was done after this if branch before, so now we only want to do the Append when the if branch doesn't trigger.
Is that correct?

Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs]: Yes. I'm confident that is what is happening. What an elegant solution.

File: src/Build/Evaluation/Expander.cs [hunk 2]
```diff
@@ -1033,94 +1052,94 @@ internal static string ExpandMetadataLeaveEscaped(string expression, IMetadataTa
...
+                ErrorUtilities.VerifyThrow(itemMetadataMatch.Success, "Need a valid item metadata.");
 
-                private LoggingContext _loggingContext;
+                string metadataName = itemMetadataMatch.Groups[RegularExpressions.NameGroup].Value;
+                string itemType = null;
 
-                /// <summary>
-                /// Constructor taking a source of metadata.
-                /// </summary>
-                internal MetadataMatchEvaluator(
```
Comment (SimaTian) [inline src/Build/Evaluation/Expander.cs:1061]: From what I read, the underlying logic is mostly the same, however for some reason the Git is quite confused and marked large portions of the code as new/replaced.
I think this is due to the constructor being extracted which caused lot of code to be shifted by one indentation level.
Weird to parse, but overall it looks fine.

File: src/Build/Evaluation/Expander.cs [hunk 3]
```diff
@@ -3171,6 +3190,107 @@ private static partial class RegularExpressions
             /**************************************************************************************************************************
              * WARNING: The regular expressions above MUST be kept in sync with the expressions in the ProjectWriter class.
              *************************************************************************************************************************/
+
+            public static void ReplaceAndAppend(string input, Func<Match, MetadataMatchEvaluator, string> evaluator, MetadataMatchEvaluator metadataMatchEvaluator, SpanBasedStringBuilder builder, Regex regex)
+            {
+                ReplaceAndAppend(input, evaluator, metadataMatchEvaluator, -1, regex.RightToLeft ? input.Length : 0, builder, regex);
```
Comment (SimaTian) [inline src/Build/Evaluation/Expander.cs]: is there ever case in our code where we use the regex.RightToLeft?
It's most likely good to have it covered, but I can't help but wonder if there are ever any metadata that would use it.

**Review by SimaTian — APPROVED** [review]
I like it and I think it is safe enough.
The expander is well covered for a lot of edge cases, see for example Expander_Tests.cs -> Medley.
Previously when I was playing around in expander, it caught basically anything so if the tests are passing, we should be good to go.

File: src/Build/Evaluation/Expander.cs [hunk 4]
```diff
@@ -3171,6 +3190,107 @@ private static partial class RegularExpressions
...
+            }
+
+            public static void ReplaceAndAppend(string input, Func<Match, MetadataMatchEvaluator, string> evaluator, MetadataMatchEvaluator metadataMatchEvaluator, int count, int startat, SpanBasedStringBuilder stringBuilder, Regex regex)
```
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs]: please add a documentation to the public methods to explain briefly intent.

File: src/Build/Evaluation/Expander.cs [hunk 5]
```diff
@@ -3171,6 +3190,107 @@ private static partial class RegularExpressions
...
+                        if (--count == 0)
+                        {
+                            break;
+                        }
+
+                        match = match.NextMatch();
+                    }
+                    while (match.Success);
+
+                    if (num2 > 0)
```
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs]: could you please change the naming for num,num2 and num3 - it's difficult to follow.

File: src/Build/Evaluation/Expander.cs [hunk 6]
```diff
@@ -3171,6 +3190,107 @@ private static partial class RegularExpressions
...
+                }
+
+                if (count == 0)
+                {
+                    stringBuilder.Append(input);
+
+                    return;
+                }
+
+                Match match = regex.Match(input, startat);
```
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs:3528]: do we need nullcheck for regex?

File: src/Build/Evaluation/Expander.cs (hunk 1 above)
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs:1015]: it also confuses me now - what is in `vectorExpression` now?

Comment (YuliiaKovalova) [reply src/Build/Evaluation/Expander.cs]: I assume we don't need `vectorExpression` anymore and can use itemVectorExpressions[n].Value directly here.

File: src/Build/Evaluation/Expander.cs [hunk 7]
```diff
@@ -972,10 +987,10 @@ internal static string ExpandMetadataLeaveEscaped(string expression, IMetadataTa
                         using SpanBasedStringBuilder finalResultBuilder = Strings.GetSpanBasedStringBuilder();
 
                         int start = 0;
-                        MetadataMatchEvaluator matchEvaluator = new MetadataMatchEvaluator(metadata, options, elementLocation, loggingContext);
```
Comment (YuliiaKovalova) [inline src/Build/Evaluation/Expander.cs:976]: could you please explain why you moved this instantiation in the `if`s?

**Comment by YuliiaKovalova** [comment]
Overall looks good, `ReplaceAndAppend` is undocumented and slightly challenging to follow. Please adjust it

File: src/Build/Evaluation/Expander.cs (hunk 3 above)
Comment (Erarndt) [reply src/Build/Evaluation/Expander.cs]: Not that I ran across. It was easy to copy over, so I did it for the sake of completeness. I can remove it if we're confident it'll never be used.

File: src/Build/Evaluation/Expander.cs (hunk 7 above)
Comment (Erarndt) [reply src/Build/Evaluation/Expander.cs:976]: This is an artifact from before I switched `MetadatamatchEvaluator` to be a struct and to pass it into `ReplaceAndAppend`. The motivation behind moving it was to avoid allocating it in unnecessary paths such as when `itemVectorExpressions` was null or empty. The benefit isn't as pronounced now that this is a struct. I can move it back if the team feels like it reads better that way.

File: src/Build/Evaluation/Expander.cs (hunk 1 above)
Comment (Erarndt) [reply src/Build/Evaluation/Expander.cs]: I've cleaned this up a bit. Let me know if there's something else you'd like to see changed.

File: src/Build/Evaluation/Expander.cs (hunk 2 above)
Comment (Erarndt) [reply src/Build/Evaluation/Expander.cs:1061]: Yeah, the logic should be the same. The major difference is that we can avoid repeatedly creating the delegate for `ExpandSingleMetadata`. Since this function is static, the compiler will cache the delegate.

File: src/Build/Evaluation/Expander.cs (hunk 4 above)
Comment (Erarndt) [reply src/Build/Evaluation/Expander.cs]: Added. Let me know if you would like to see any changes.

File: src/Build/Evaluation/Expander.cs (hunk 6 above)
Comment (Erarndt) [reply src/Build/Evaluation/Expander.cs:3528]: Added!

File: src/Build/Evaluation/Expander.cs (hunk 5 above)
Comment (Erarndt) [reply src/Build/Evaluation/Expander.cs]: Updated naming to match the actual names in source. I was looking at the decompiled code before which hadn't preserved the names.

**Review by YuliiaKovalova — APPROVED** [review]

File: src/Build/Evaluation/Expander.cs (hunk 3 above)
Comment (SimaTian) [reply src/Build/Evaluation/Expander.cs]: In that case I think it's fine.
